zero/src/rpc/jumpdest.rs

```diff
@@ -0,0 +1,516 @@
+
+    // This map does not contain `initcodes`.
+    let callee_addr_to_code_hash: HashMap<Address, H256> = tx_traces
+        .iter()
+        .map(|(callee_addr, trace)| (callee_addr, &trace.code_usage))
+        .filter(|(_callee_addr, code_usage)| code_usage.is_some())
+        .map(|(callee_addr, code_usage)| {
+            (*callee_addr, code_usage.as_ref().unwrap().get_code_hash())
+        })
+        .collect();
+
+    trace!(
+        "Transaction: {} is a {}.",
+        tx.hash,
+        if tx.to.is_some() {
+            "message call"
+        } else {
+            "contract creation"
+        }
+    );
+
+    let entrypoint_code_hash: H256 = match tx.to {
+        Some(to_address) if precompiles().contains(&to_address) => return Ok(jumpdest_table),
+        Some(to_address) if callee_addr_to_code_hash.contains_key(&to_address).not() => {
+            return Ok(jumpdest_table)
+        }
+        Some(to_address) => callee_addr_to_code_hash[&to_address],
+        None => {
+            let init = &tx.input;
+            keccak(init)
+        }
+    };
+
+    // `None` encodes that previous `entry` was not a JUMP or JUMPI with true
+    // condition, `Some(jump_target)` encodes we came from a JUMP or JUMPI with
+    // true condition and target `jump_target`.
+    let mut prev_jump: Option<U256> = None;
+
+    // Call depth of the previous `entry`. We initialize to 0 as this compares
+    // smaller to 1.
+    //let mut prev_depth = 0;
+    // The next available context. Starts at 1. Never decrements.
+    let mut next_ctx_available = 1;
+    // Immediately use context 1;
+    let mut call_stack = vec![(entrypoint_code_hash, next_ctx_available)];
+    next_ctx_available += 1;
+
+    for (step, entry) in struct_log.iter().enumerate() {
+        let op = entry.op.as_str();
+        let curr_depth: usize = entry.depth.try_into().unwrap();
+
+        ensure!(curr_depth <= next_ctx_available, "Structlog is malformed.");
+
+        while curr_depth < call_stack.len() {
+            call_stack.pop();
+        }
+
+        ensure!(
+            call_stack.is_empty().not(),
+            "Call stack was unexpectedly empty."
+        );
+        let (code_hash, ctx) = call_stack.last().unwrap();
+
+        trace!("TX:   {:?}", tx.hash);
+        trace!("STEP: {:?}", step);
+        trace!("STEPS: {:?}", struct_log.len());
+        trace!("OPCODE: {}", entry.op.as_str());
+        trace!("CODE: {:?}", code_hash);
+        trace!("CTX:  {:?}", ctx);
+        trace!("CURR_DEPTH:  {:?}", curr_depth);
+        trace!("{:#?}\n", entry);
+
+        match op {
+            "CALL" | "CALLCODE" | "DELEGATECALL" | "STATICCALL" => {
+                prev_jump = None;
+                ensure!(entry.stack.as_ref().is_some(), "No evm stack found.");
+                // We reverse the stack, so the order matches our assembly code.
+                let evm_stack: Vec<_> = entry.stack.as_ref().unwrap().iter().rev().collect();
```

Comment:
This is called 4 times, I think we could do it just once ahead for clarity before the match statement

Reply:
It is only called once per `StructLog` entry. The reason I don't just do it once, and for all `op`s, is that for most `op`s, we do not need to do anything, so reversing the stack would be wasteful.